src/Microsoft.AspNet.Razor.Runtime/TagHelpers/DefaultTagHelperContent.cs

```diff
@@ -158,66 +141,119 @@ public override TagHelperContent AppendFormat([NotNull] string format, params ob
-                {
-                    Append(value);
-                }
-            }
-
-            // If Append() was called with an empty TagHelperContent IsModified should
-            // still be true. If the content was not already modified, it means it is empty.
-            // So the Clear() method can be used to indirectly set the IsModified.
-            if (!IsModified)
-            {
-                Clear();
-            }
-
+            Buffer.Append(htmlContent);
             return this;
         }
 
         /// <inheritdoc />
         public override TagHelperContent Clear()
         {
-            _buffer.Clear();
+            Buffer.Clear();
             return this;
         }
 
         /// <inheritdoc />
         public override string GetContent()
         {
-            return string.Join(string.Empty, _buffer);
+            if (_buffer == null)
+            {
+                return string.Empty;
+            }
+
+            return string.Join(string.Empty, Buffer);
         }
 
         /// <inheritdoc />
         public override string ToString()
         {
-            return GetContent();
+            using (var writer = new StringWriter())
+            {
+                WriteTo(writer, new HtmlEncoder());
+                return writer.ToString();
+            }
         }
 
         /// <inheritdoc />
-        public override IEnumerator<string> GetEnumerator()
+        public override void WriteTo([NotNull] TextWriter writer, [NotNull] IHtmlEncoder encoder)
         {
-            // The enumerator is exposed so that SetContent(TagHelperContent) and Append(TagHelperContent)
-            // can use this to iterate through the values of the buffer.
-            return _buffer.GetEnumerator();
+            Buffer.WriteTo(writer, encoder);
         }
 
-        /// <inheritdoc />
-        public override void WriteTo(TextWriter writer, IHtmlEncoder encoder)
+        private bool CheckIfBufferedHtmlContentIsEmpty(BufferedHtmlContent buffer)
         {
-            foreach (var entry in _buffer)
+            foreach (var value in buffer)
             {
-                writer.Write(entry);
+                var valueAsString = value as string;
+                if (valueAsString != null)
+                {
+                    if (!string.IsNullOrEmpty(valueAsString))
+                    {
+                        return false;
+                    }
+                }
+                else
+                {
+                    var valueAsBufferedHtmlContent = value as BufferedHtmlContent;
+                    if (valueAsBufferedHtmlContent != null)
+                    {
+                        if (!CheckIfBufferedHtmlContentIsEmpty(valueAsBufferedHtmlContent))
+                        {
+                            return false;
+                        }
+                    }
+                    else
+                    {
+                        if (!string.IsNullOrEmpty(value.ToString()))
```

Comment:
Does this correctly handle MVC's (distinct) `BufferedHtmlContent` class, `HtmlString` or `StringHtmlContent`? Further nothing in the `IHtmlContent` interface guarantees an implementation has overridden `ToString()` in a manner that's useful here. Probably need to add something like `GetRawString()` to that interface.